cpp/src/arrow/filesystem/path_util.cc

```diff
@@ -287,6 +288,38 @@ bool IsLikelyUri(util::string_view v) {
   return ::arrow::internal::IsValidUriScheme(v.substr(0, pos));
 }
 
+struct Globber::Impl {
+  std::regex pattern_;
+
+  explicit Impl(std::string p) : pattern_(std::regex(std::move(transform(p)))) {}
+
+  std::string transform(std::string p) {
```

Comment:
I suspect there are other regex characters to worry about.  For example, in a URI path you are allowed to use `(` and `)`.

If we are going to go this route we either should make it very clear somewhere what characters we expect to see in a filename or we need to escape all of `.^$*+?()[{\|`

Pity there is no equivalent to https://docs.python.org/3/library/re.html#re.escape

Reply:
Yes, we can simply escape all of that. A switch/case statement would work.

Reply:
Also, can reuse the same set of characters as Python:
https://github.com/python/cpython/blob/37c6db60f9ac62b8a80bf04a8146274756ee0da0/Lib/re/__init__.py#L232-L247

Reply:
I have a question about * ? wildcards, how we should treat them in remote filesystems? I think we can an escape in a previous step or say somewhere that for this wildcards is needed to add an escape if they are not used as wildcards: `s3://mybucket/foo/bar/?region=utopia&endpoint_override=localhost&scheme=http`

Reply:
Looks like POSIX allows for escaping special characters in a glob:
```
One can remove the special meaning of '?', '*', and '[' by
       preceding them by a backslash, or, in case this is part of a
       shell command line, enclosing them in quotes.
```

So a user could do...
```
s3://mybucket/foo/bar/\?region=utopia*
```

That being said...the S3 filesystem is going to be matching against bucket/key only I think.  So the particular URL you listed isn't going to work as a glob expression anyways.

Escaping is a concern though.  `*` is a legal character in an S3 key.

Reply:
> I think we can an escape in a previous step or say somewhere that for this wildcards is needed to add an escape if they are not used as wildcards: `s3://mybucket/foo/bar/?region=utopia&endpoint_override=localhost&scheme=http`

We are working on paths here, not URIs. The path in your example would be `mybucket/foo.bar`

Reply:
Ok, got it

Reply:
I think escaping these characters could still potentially be a concern.  `*` and `?` are legal filename characters on Linux and S3.

s3://weston-my-bucket/*
https://weston-my-bucket.s3.amazonaws.com/*

I'm fine with leaving that for a follow-up however.

Reply:
I think we can escape both of them. These paths should be something like:
s3://weston-my-bucket/\*
https://weston-my-bucket.s3.amazonaws.com/\*

So, * and ? won't be interpreted as a wildcard

Reply:
Well, the burden of escaping the path would initially be on the user.  For example, in S3 I could have the following paths:
```
weston-my-bucket/*/0.parquet
weston-my-bucket/*/1.parquet
```

As a user I might pass in the glob URI: `s3://weston-my-bucket/\*/*.parquet`

We then need to make sure that the regex we generate is (I think): `weston-my-bucket/\*/.*\.parquet`

Reply:
On the other hand if the user provides the glob URI: `s3://weston-my-bucket/*` then we use the regex `weston-my-bucket/.*`

Reply:
Ok, I think I can make that change